### ISSUE TITLE
Add a "RichTextLabel3D" node

### ISSUE THREAD

**Issue by bhiltune** [godotengine/godot-proposals#9805]
### Describe the project you are working on

Various projects that utilize floating text for stylistic effects. Walking sims, 3D-action games, etc...

For example: I want to recreate the [Fight Club Ikea scene.](https://youtu.be/exL51n3py6g?t=17) That's currently not possible to do in an intuitive, fast manner.

### Describe the problem or limitation you are having in your project

There is currently no way to display RichTextLabel-text in 3D-space without convoluted workarounds. Many ordinary stylistic choices are thus unreasonably hard to implement, such as text displays on museum/gallery walls.

![museum-exhibition-labels-deyoung-1](https://github.com/godotengine/godot-proposals/assets/52609762/71a410fe-d498-402e-8485-305d54136080)

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

Add a way to quickly create Label3D nodes with Rich Text functionailty, like **bold text**, _italics_, BBCode.

From the Godot Docs on Label(2D) nodes:

"[Label](https://docs.godotengine.org/en/stable/classes/class_label.html#class-label) nodes are great for displaying basic text, but they have limitations. If you want to change the color of the text, or its alignment, you can only do that to the entire label. You can't make a part of the text have another color, or have a part of the text centered. To get around these limitations, you would use a [RichTextLabel](https://docs.godotengine.org/en/stable/classes/class_richtextlabel.html#class-richtextlabel).

[RichTextLabel](https://docs.godotengine.org/en/stable/classes/class_richtextlabel.html#class-richtextlabel) allows for complex formatting of text using a markup syntax or the built-in API. It uses BBCodes for the markup syntax, a system of tags that designate formatting rules for a part of the text. You may be familiar with them if you ever used forums (also known as bulletin boards, hence the "BB" in "BBCode")..."

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

Presumably the same principles could be applied as what's going on between a Label and a Label3D, just with the RichTextLabel.

### If this enhancement will not be used often, can it be worked around with a few lines of script?

Not really - most solutions require the utilization of Viewports and ViewportTextures, which is time-intensive and non-intuitive compared to the simplicity of having a dedicated node, like Label3D.

While this isn't something absolutely necessary for most people, the same could technically be said about the RichTextLabel node.

### Is there a reason why this should be core and not an add-on in the asset library?

I think it would be consistent with the rest of the engine to have this sort of node. We currently have equivalents for almost every node between 2D and 3D workflows, so transferring knowledge of the engine from one to the other is very intuitive and simple. 

Adding this node would serve to make the two workflows even more similar and fix an annoying "hole" in the 3D capabilities of the engine that currently has to be worked around.

**Comment by passivestar** [comment]
I really believe godot needs a proper 3D UI system that would make all controls somehow support 3D, implementing them on per-node basis one by one doesn't seem right because it will introduce a lot of code duplication making it hard to maintain. These proposals will keep popping up for every single control node out there https://github.com/godotengine/godot-proposals/discussions/4093

**Comment by AThousandShips** [comment]
Indeed I don't think we should duplicate all controls for 3D but instead improve usability of 2D in 3D and make input easier for that

That being said using interfaces that are *actually* in the 3D space is usually less than perfect even when implemented well, and largely gimmicky, with the exception of well used XR implementations, which is at least partially a case of its own

**Comment by passivestar** [comment]
> but instead improve usability of 2D in 3D

my only concern is that a UI system that doesn't allow full per-node 3D and only lets UI elements live in a pre-defined 3D "containers" may be a half-measure that would have to be revisited in the future when people realize its limitations. Compare it to Unity where you can for example _animate in_ a big grid of UI elements that all have their own 3d rotation offset and so on

I understand where you're coming from calling it gimmicky because it's very easy to overdo and also UI fashion and the sense of what's cool or not changes from year to year. There's a line between making something cringe and something subtle, and it should be up to a UI designer to make those choices. The underlying technology should allow you to make things that look cringe too if that's what you're going for. Games are the best medium for creative expression and we can't predict how people will use it, that's the beauty of it

XR may be its own niche now and it will likely take a long time for technology to catch up before it'll change the way we interact with 3d worlds (a lot of hardware problems will have to be solved), but more likely than not this time will come

I think a system that's less opinionated, covers more potential use-cases and gives full creative control to the user is a good goal to have, that's why I'm personally for switching the UI system to 3D fully like in Unity, but with a better UX that will make 3D aspect of it "invisible" until you need it

**Comment by Gnumaru** [comment]
> There is currently no way to display RichTextLabel-text in 3D-space without convoluted workarounds

As bhiltune said, there are "convoluted workarounds".

![image](https://github.com/godotengine/godot-proposals/assets/5783414/b40a40a0-04fa-42b7-aef0-3e06422d8376)

> I really believe godot needs a proper 3D UI system that would make all controls somehow support 3D

As passivestar said, we could have some simple system to make every control node "just work" on 3D.

Maybe we could have some sort of "Canvas3D" that is just a facade, a Node3D that, under the hood, just put all its children under a subviewport, renders it on a quad using a viewporttexture, configure all through some proxy properties and call it a day. But all this could be done as an addon, not necessarily a core node type.

**Comment by bhiltune** [comment]
I feel like some of the conversation has shifted to specifically UX and User Interaction in 3D space, which are important considerations but the original suggestion stemmed from simply wanting to display rich text in a 3D space, a feature that I think is pretty reasonable for the engine to offer in a simple manner. As Passivestar mentioned, the engine should be built without any notions of what design decisions the dev _should_ make, but rather allow any number of combinations from predefined blocks. Lots of innovation can arise from that!

And yeah, "convoluted" may definitely have quotes around it, but I am of the opinion that having this type of feature gated behind other engine concepts hurts workflow and discovery when compared to having a dedicated node. It also presents a variety of issues like worse organization of scenes, unnecessary complications, harder adoption for new users...

I do really like the idea of streamlining the interaction between UI elements and 3D space, and feel like there's some good points about having some kinda "3D toggle" or "Canvas3D" node. It feels like a pretty elegant solution that would avoid too many new nodes while allowing for both this and any number of similar requests to be rapidly implemented.

**Comment by HKunogi** [comment]
A quick and simple example of that being useful, while making a card game in 3D world, in which cards have too many details and specifics on the design, which the card texts are complex and requires actual BBCode on it (for things like bold, italic and such), it would be magnificent to be able to simple have a RichTextLabel3D, with scrolling support and all that on 3D.

**Comment by morphles** [comment]
This is indeed separate from 3D UI (which with VR/AR is also something that might be needed, but likely too early to have something built in specifically as there probably are less conventions/obvious things, and developers should just cobble stuff up using regular 3d stuff?), since as others mentioned, rich text label is just that - text. No need for it to be interactive. For now I do not even find any examples of how to do that "somewhat sanely" and scalably. Above viewport example I guess is of you just need one. But AFAIK viewports are not that great for performance. So... not super usable for me. Some parts, I can conceive would be good to pre-render into texture on game load. But it is not clear how to do that, and I do not find examples.

**Comment by Calinou** [comment]
>  Some parts, I can conceive would be good to pre-render into texture on game load.

This is what Viewport can do if you set its update mode to **Once** :slightly_smiling_face: 

The performance overhead is minimal in this case.

**Comment by beoceafy-inc** [comment]
Godot need to add RichTextLabel3D for 3D space becouse im making a 3D game multiplayer and i want make 3D texts for the player nickname like in Minecraft or Stumble Guys but i want use RichTextLabel BBCode feature witch does not have in Label3D

**Comment by LemmaEOF** [comment]
Yeah, it really doesn't feel like you should have to dedicate an entire framebuffer to rendering a piece of rich text in a 3D environment. We already have a `Label3D` class that's distinct from the `Control` hierarchy, so a `RichTextLabel3D` would make plenty of sense. It could open the way for 3D text modifiers too - moving along the Z-axis or the like.

**Comment by PauloVBettio** [comment]
The problem I see with using viewports for 3D Text is not performance only, is that they require you to manually set the resolution and, of course, you lose a lot of 3D features that exists on Label3D (such as outline ordering).

There should be at least a simple solution for formatting Label3D texts into having Italics and Bold spans of text without having to rely on prerendering.

How does other engines handle 3D texts? I remember unity used to have a 3d text with rich text features, I don't know how they are now.

**Comment by BednarStudios** [comment]
i found an godot asset called "RichText3D" that its the same RichTextLabel but in 3D, just search RichText3D on asset library

**Comment by LemmaEOF** [comment]
the RichText3D plugin does exist, but as I said it requires you to dedicate a whole framebuffer for every single piece of text on screen and doesn't support 3D-space text modifiers.

**Comment by BednarStudios** [comment]
> the RichText3D plugin does exist, but as I said it requires you to dedicate a whole framebuffer for every single piece of text on screen and doesn't support 3D-space text modifiers.

here is the asset link: [RichText3D](https://godotengine.org/asset-library/asset/4291)

**Comment by LemmaEOF** [comment]
Yes, I looked the library up and checked the source code. [It allocates a framebuffer for every piece of text](https://github.com/mszylkowski/rich-text-3d-godot/blob/main/addons/rich_text_3d/rich_text_3d.gd#L52). A SubViewport (Godot's system for allocating framebuffers) per node is far too high of a cost if you want to have, say, thirty to fifty pieces of text floating around on-screen.

**Comment by Gnumaru** [comment]
The SubViewport node has a render_target_update_mode property. Unless I'm missunderstood, If you set it to SubViewport.UpdateMode.UPDATE_ONCE it wil only render a single frame and after that it will reuse that same data, so I think it will have effectively the same computational cost of rendering a static texture of the same size.

https://docs.godotengine.org/en/latest/classes/class_subviewport.html#enum-subviewport-updatemode

BUT you won't be able to use animated effects.

**Comment by LemmaEOF** [comment]
Yeah, the animated effects are the real problem - that also puts a limit on text scrolling out over time. A more elegant solution would be much preferred if possible.
